### PR TITLE
Add the `force_latest_compatible_version` input, and add the "auto-detect Dependabot/CompatHelper" functionality

### PR DESCRIPTION
**Pull request by DilumAluthge**
This should not be merged until https://github.com/JuliaLang/Pkg.jl/pull/2439 is merged.

### CODE REVIEW

File: action.yml
Comment (christopher-dG) [inline action.yml:39]: This should probably be put on multiple lines

Comment (DilumAluthge) [reply action.yml:39]: Sure! I'm not sure what the syntax for that is.

Comment (DilumAluthge) [reply action.yml:39]: @christopher-dG Is this correct?
```yaml
- run: |
  julia --color=yes --check-bounds=yes --inline=${{ inputs.inline }} --project -e '
using Pkg;
kwargs = Dict{Symbol, Any}();
kwargs[:coverage] = ${{ inputs.coverage }; ${{ inputs.force_latest_compat }} && (kwargs[:force_latest_compat] = ${{ inputs.force_latest_compat }});
Pkg.test(; kwargs...)'
  shell: bash
```

Comment (christopher-dG) [reply action.yml:39]: That looks good. I think you don't even need the semicolons.

Comment (christopher-dG) [reply action.yml:39]: Except do indent the string by two more spaces.

Comment (DilumAluthge) [reply action.yml:39]: Alright, how about this:
```yaml
- run: |
  julia --color=yes --check-bounds=yes --inline=${{ inputs.inline }} --project -e '
    using Pkg
    kwargs = Dict{Symbol, Any}()
    kwargs[:coverage] = ${{ inputs.coverage }
    ${{ inputs.force_latest_compat }} && (kwargs[:force_latest_compat] = ${{ inputs.force_latest_compat }})
    Pkg.test(; kwargs...)'
  shell: bash
```

Comment (christopher-dG) [reply action.yml:39]: The whole thing needs to be indented past the parent `run`, so:

```yml
- run: |
    julia ... -e '
      using Pkg ...'
```

Comment (DilumAluthge) [reply action.yml:39]: Like this?

```yaml
- run: |
    julia --color=yes --check-bounds=yes --inline=${{ inputs.inline }} --project -e '
      using Pkg
      kwargs = Dict{Symbol, Any}()
      kwargs[:coverage] = ${{ inputs.coverage }
      ${{ inputs.force_latest_compat }} && (kwargs[:force_latest_compat] = ${{ inputs.force_latest_compat }})
      Pkg.test(; kwargs...)'
  shell: bash
```

Comment (christopher-dG) [reply action.yml:39]: Yep that parses. This seems like the dumbest nitpick ever but YAML really is this picky 😅

Comment (DilumAluthge) [reply action.yml]: Third time's the charm 😂 

Yeah this YAML stuff always kills me.

I've updated this PR - can you double check the syntax one last time?

Also, I've updated https://github.com/julia-actions/composite-runtest/pull/1 with the same change - can you check the syntax there as well?

Comment (christopher-dG) [reply action.yml]: Both look good to me.

**Comment by christopher-dG** [comment]
Maybe there should be a version check if the feature is turned on? If we're less than 1.7, print a warning or something.

**Comment by DilumAluthge** [comment]
> Maybe there should be a version check if the feature is turned on? If we're less than 1.7, print a warning or something.

Good idea! Take a look at the commit I just pushed - does that meet the requirement?

**Comment by DilumAluthge** [comment]
If it looks good, I'll add it to the other PR as well.

File: action.yml
Comment (SaschaMann) [inline action.yml:43]:
```suggestion
          kwargs[:coverage] = ${{ inputs.coverage }}
```

File: action.yml
Comment (SaschaMann) [inline action.yml:45]:
```suggestion
              if VERSION >= v"1.7.0-" # excludes 1.6, includes 1.7-DEV, includes 1.7
```

I think this is equivalent but slightly easier to read:

```julia
julia> v = v"1.7.0-"
v"1.7.0-"

julia> v"1.6.4294967295" >= v
false

julia> v"1.7-DEV" >= v
true

julia> v"1.7.0-" >= v
true

julia> v"1.7.0-DEV" >= v
true

julia> v"1.7.0-alpha" >= v
true

julia> v"1.7.0-rc1" >= v
true

julia> v"1.7.0-0" >= v
true
```

**Review by SaschaMann — CHANGES_REQUESTED** [review]

**Comment by SaschaMann** [comment]
I wonder if instead of adding individual inputs for `Pkg.test(;kwargs...)` we should take a single input for all kwargs? Or add inputs for every kwarg?

**Comment by DilumAluthge** [comment]
> I wonder if instead of adding individual inputs for `Pkg.test(;kwargs...)` we should take a single input for all kwargs? Or add inputs for every kwarg?

I don't feel strongly either way. What do you think users would prefer?

**Comment by DilumAluthge** [comment]
> I wonder if instead of adding individual inputs for `Pkg.test(;kwargs...)` we should take a single input for all kwargs? Or add inputs for every kwarg?

So, we've already added individual inputs for:
- `coverage`
- `depwarn`
- `inline`

So I think probably it would be easiest to just keep the convention of having individual inputs?

**Comment by SaschaMann** [comment]
Sounds good. 

Changes look good to me, I'll merge when the JuliaLang/Pkg.jl#2176 is merged (and leave it on change request until then to prevent accidental merges)

File: action.yml
Comment (christopher-dG) [inline action.yml:51]:
```suggestion
                  @warn("The `force_latest_compat` option requires at least Julia 1.7", VERSION)
```

**Review by christopher-dG — APPROVED** [review]

**Comment by DilumAluthge** [comment]
I've added the "auto-detect Dependabot/CompatHelper" functionality from https://github.com/JuliaLang/Pkg.jl/pull/2234

@christopher-dG @SaschaMann Can you review again?

File: action.yml
Comment (SaschaMann) [inline action.yml:17]:
```suggestion
    description: 'If true, then, for each `[compat]` entry in the active project, only the latest compatible version will be allowed. Options: true | false | autodetect. Default value: auto. If the value is auto, then force_latest_compat will be set to true if this is a Dependabot/CompatHelper pull request, and false otherwise.'
    default: 'auto'
```

File: autodetect-dependabot.jl
Comment (SaschaMann) [inline autodetect-dependabot.jl:18]: Could this be done with `replace` instead? I feel like that might be nicer than hard-coding the length.

File: autodetect-dependabot.jl
Comment (christopher-dG) [inline autodetect-dependabot.jl:3]: None of this is necessary if it's only for GHA.

File: autodetect-dependabot.jl
Comment (SaschaMann) [inline autodetect-dependabot.jl:25]: I think we might want this to be `dependabot/julia` (guessing based on https://github.com/julia-actions/setup-julia/pull/47), so that it doesn't affect non-Julia Dependabot PRs in mixed repos.

File: autodetect-dependabot.jl
Comment (DilumAluthge) [reply autodetect-dependabot.jl:3]: Good catch. I'll fix that now.

File: action.yml
Comment (SaschaMann) [inline action.yml:16]:
```suggestion
    description: 'If true, then, for each `[compat]` entry in the active project, only the latest compatible version will be allowed. Options: true | false | auto. Default value: auto. If the value is auto, then force_latest_compat will be set to true if this is a Dependabot/CompatHelper pull request, and false otherwise.'
```

Sorry, I missed one.

File: autodetect-dependabot.jl
Comment (DilumAluthge) [reply autodetect-dependabot.jl:3]: Alright I've removed the Travis and AppVeyor stuff.

File: autodetect-dependabot.jl
Comment (DilumAluthge) [reply autodetect-dependabot.jl:21]: @christopher-dG Can we ensure that Julia Dependabot PRs start with `dependabot/julia`?

Comment (christopher-dG) [reply autodetect-dependabot.jl:21]: Yes

File: autodetect-dependabot.jl
Comment (DilumAluthge) [reply autodetect-dependabot.jl]: Done.

File: autodetect-dependabot.jl
Comment (DilumAluthge) [reply autodetect-dependabot.jl]: Done!

**Comment by SaschaMann** [comment]
Should this be considered a breaking change and require a major version bump of the action?

**Comment by DilumAluthge** [comment]
Hmmm. The only change in behavior is on CompatHelper/Dependabot PRs, right?

Dependabot doesn't exist yet.

And for CompatHelper, this was always supposed to be the intended behavior, so honestly I would consider it to be a bug fix.

File: autodetect-dependabot.jl
Comment (SaschaMann) [inline autodetect-dependabot.jl:14]: For the (very unlikely) case that someone names their branch `refs/heads`, let's only replace the first occurence

```suggestion
        return replace(branch_name, "refs/heads/" => "", count = 1)
```

**Comment by DilumAluthge** [comment]
As long as the behavior change is only on Dependabot/CompatHelper PRs, I would say this "technically is a breaking change in behavior, but really more of a bug fix, so it's okay to do it in a patch release".

**Comment by DilumAluthge** [comment]
> the behavior change is only on Dependabot/CompatHelper PRs

As long as this is true.

File: autodetect-dependabot.jl
Comment (SaschaMann) [inline autodetect-dependabot.jl:14]: For the (very unlikely) case that someone names their branch `refs/heads`, let's only replace the first occurence and only if it's at the start of the string:

```suggestion
        return replace(branch_name, r"^(refs\/heads\/)" => "")
```

**Comment by DilumAluthge** [comment]
I've added another small change: if the user sets `force_latest_compat` to `auto` (the default value), and this is in fact a Dependabot/CompatHelper job, then I print an `@info` message saying:
> @info("This is a Dependabot/CompatHelper job, so `force_latest_compat` has been set to `true`")

That way, if a user looks at the logs, they'll know that `force_latest_compat` was set to be `true`.

**Comment by SaschaMann** [comment]
Looks good to me :)

**Comment by DilumAluthge** [comment]
Once https://github.com/JuliaLang/Pkg.jl/pull/2439 is merged, I'll fix the merge conflicts and rebase this PR on the latest master.

**Comment by DilumAluthge** [comment]
@christopher-dG @SaschaMann We are still waiting on https://github.com/JuliaLang/Pkg.jl/pull/2439 to be merged, but in the mean time, I have rebased and fixed the merge conflicts. Can you review this again?

**Comment by DilumAluthge** [comment]
We'll also need to wait until:
1. https://github.com/JuliaLang/julia/pull/40266 is merged into master
2. The nightly binaries are built and unloaded from that commit on master

**Comment by DilumAluthge** [comment]
Okay, https://github.com/JuliaLang/julia/pull/40266 has been merged.

**Comment by DilumAluthge** [comment]
FYI the macOS Buildbots are currently broken.

File: action.yml
Comment (SaschaMann) [inline action.yml:17]:
```suggestion
    description: 'If true, then, for each [compat] entry in the active project, only allow the latest compatible version. If the value is auto and the pull request has been opened by Dependabot or CompatHelper, then force_latest_compatible_version will be set to true, otherwise it will be set to false. Options: true | false | auto. Default value: auto.'
```

File: action.yml
Comment (SaschaMann) [inline action.yml:63]: Why is this split into two steps?

File: kwargs.jl
Comment (SaschaMann) [inline kwargs.jl:21]:
```suggestion
            is_dependabot_job && @info("This is a Dependabot/CompatHelper job, so `force_latest_compatible_version` has been set to `true`")
            kwargs_dict[:force_latest_compatible_version] = is_dependabot_job
```

(Untested and written in the GH suggestion prompt)

Remove another nesting level here will also help make understanding what's going on easier, I think.

File: kwargs.jl
Comment (SaschaMann) [inline kwargs.jl:23]: Could you move this to a guard clause all the way at the top of the function, please?

File: autodetect-dependabot.jl
Comment (SaschaMann) [inline autodetect-dependabot.jl:17]:
```suggestion
    replace(branch_name, r"^(refs\/heads\/)" => "")
```

If the regex doesn't match anything, nothing will be replaced. We don't need to check for that case manually.

**Review by SaschaMann — COMMENTED** [review]
Thanks for keeping this up-to-date!

I found it a bit hard to follow what's going on in the script. I left a few comments on how, imo, that could be improved, mostly by removing nesting levels.

Happy to apply these changes myself in a follow-up PR too

File: kwargs.jl
Comment (DilumAluthge) [reply kwargs.jl:11]: macOS buildbots have been broken for approximately a week, so the nightlies  for macOS are out of date. A PR was just merged that will fix this, so soon the macOS nightlies will be up-to-date, and we can remove this before merging this PR.

File: kwargs.jl
Comment (DilumAluthge) [reply kwargs.jl]: Done

File: kwargs.jl
Comment (DilumAluthge) [reply kwargs.jl]: Done

**Comment by DilumAluthge** [comment]
@SaschaMann Take a look now.

File: action.yml
Comment (DilumAluthge) [reply action.yml:63]: Would I just change `${{ env.JULIA_CMD }}` to `"$JULIA_CMD"`

**Comment by DilumAluthge** [comment]
We'll need to wait to merge until the macOS nightlies are fixed.

**Comment by DilumAluthge** [comment]
Alright the macOS buildbots are fixed and the macOS nightlies are up to date.

This is good to merge from my end.

@SaschaMann

File: action.yml
Comment (SaschaMann) [reply action.yml:63]: You could leave the bash part almost as it was and then only replace `-e 'using Pkg; Pkg.test(coverage=${{ inputs.coverage }})'` with the new code.

I forgot the exact details, but there was some reason why `()` and a bash array was used. Originally I had a similar approach as this with normal double quoted variables.

Comment (SaschaMann) [reply action.yml:63]: Other than that, this PR is good to be merged

Comment (DilumAluthge) [reply action.yml]: @SaschaMann Take a look now. Is this correct?

Comment (SaschaMann) [reply action.yml]: Looks good to me! Thanks :)

**Review by SaschaMann — APPROVED** [review]

**Comment by DilumAluthge** [comment]
@SaschaMann Could we tag a new v1.x.y tag?

**Comment by DilumAluthge** [comment]
You beat me to it 😂 

https://github.com/julia-actions/julia-runtest/releases/tag/v1.6.0

**Comment by DilumAluthge** [comment]
@SaschaMann and @christopher-dG thank you for your help on this?

**Comment by SaschaMann** [comment]
I probably should have mentioned that I'm gonna tag it :D 

I usually do that right after merging PRs to these actions, I should probably setup a merge-and-tag script at some point.

**Comment by DilumAluthge** [comment]
🚀 🚀 🚀 it works!!! 

https://github.com/cscherrer/SossMLJ.jl/runs/2241088439 (triggered by https://github.com/cscherrer/SossMLJ.jl/pull/132)
```
┌ Error: Package is not at the latest compatible version
│   name = "CategoricalArrays"
│   uuid = UUID("324d7699-5711-5eae-9e2f-1d82baa6b597")
│   compat_entry_string = "0.8, 0.9"
│   active_version = v"0.8.3"
│   latest_compatible_version = v"0.9.4"
│   earliest_backwards_compatible_version = v"0.9.0"
│   allow_earlier_backwards_compatible_versions = true
└ @ Pkg.Operations /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.7/Pkg/src/Operations.jl:1856
ERROR: One or more direct dependencies is not at the latest compatible version
```

**Comment by DilumAluthge** [comment]
> I should probably setup a merge-and-tag script at some point.

This could be useful!

**Pull request merged**
